### DISCUSSION TITLE
Non Persistence: The Absolute Worst Aspect of GitHub IMO

### DISCUSSION THREAD

**Discussion by Myndex** [community/community#67089] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

## Bug/Feature Intro
The bug described herein seems to be getting worse and not better. It makes your web interface, especially for editing files rather unusable, and can result in loss of work.

### What
Any content being edited is not persistent if/when a page reloads, and it very much needs to be.

### Where/When
The bug/feature issues exist in their most virulent form in the **file editor** as it exists under the CODE tab in a repository. Though, it can happen anywhere an item is being edited.

Today's joyous loss of some hours of work happened on Safari 15.6.1, the most recent version of Safari I can use on this particular MacBook. I use Safari for accessibility reasons.

### Why
The bug collection has a direct negative impact on:
- accessibility
- general use-ability
- site functionality
- inability to use features (such as spellcheck)
- sudden, unrecoverable loss of work

### How
There is no persistence of content being edited in the GitHub web interface, or I perhaps should say, GitHub appears to interfere with any built-in persistence of Safari (I may be misinterpreting that however.) 

Persistence is a **_requirement_** for web based tools. If any error should happen (and it happens often on GitHub) then the user will lose all work in progress since the last commit.

_**ALL WORK IS LOST**_

And when this happens it makes me physically ill. I can't always be in the lab or studio using my desktop IDE, sometimes I want to make edits using the web interface. Sometimes, those edits begin to take significant time.

And yes, I make regular commits. But also, I often do not want to have to look through 150 commits, so commits are staggered to group changes accordingly.

But _THERE IS NO SAVE_ and _THEREFORE NO PERSISTENCE_

When there is any minor eventuality that causes a page reload, all work prior to a commit is **_lost_**.

## _Caveats and Centipedes_
Let's say the **main bug is the lack of persistence**, when any event whatsoever causes a page reload for any reason. This is not a problem on literally any other site I've used in the last five years.

### Page Reloads Caused by GitHub Crashing
There are a few ways to make the GitHub interface crash, spellcheck being one. The spellchecker in Safari does not work on GitHub's file-edit webapp, even when working on a text or Markdown document. It appears isolated to the file-edit webapp, as here for a discussion, spellcheck does indeed work.

But not only does spellcheck not work at all in file-edit, invoking the spellcheck (i.e. check document now) can **cause GitHub page to crash**, and then force reload, causing a loss of all work. There's no opportunity to freeze the reload to scavenge the work before it is forced.

The label change bug and text search bugs seem to be fixed.

### Page Reloads caused by User Click Error
I am disabled. Due to my low vision and other issues, I can easily and accidentally click someplace on the UI to cause navigation to a new page. When this happens on sites where Safari is aware of a document in progress, the Safari dialog warning works to prevent a new page from being loaded.

Apparently, GitHub bypasses this and uses their own dialog, which has no effect whatsoever.

**The GitHub dialog asks:**

<img alt="github asks a pointless question" width="420" src="https://github.com/community/community/assets/42009457/de79ecc6-45b4-4871-9793-c86f655384dc">

As you can see the GitHub dialog is there while the next page loads regardless, and clicking "stay on page" does nothing.

## Fixing Thoughts

I don't really care about the _bugs_, but I do care about losing work. 

The way to fix that is to add persistence and retention, so that if you happen to leave a particular page, when you go back at least within a certain amount of time, that your work is still there. But this is not the case throughout GirHub.

**"Orthogonal Persistence"** has been a thing in some systems for decades, and today is a thing in an increasing number of systems, apps, devices.

A pen writing on a piece of paper has inherent orthogonal persistence. The nature of work done on GitHub _demands that it does as well_. If for some reason you can't implement persistence, at least there needs to be a non-commit interim save. _PLEASE_

- Saves _should_ happen every edit, or at lease every few minutes.
- Commits should be grouped based on items/areas being worked on.
  - A commit is not a substitute for a save; it has a different purpose and function.

Thank you for reading.

### TL;DR
_Any content being edited is not persistent when a page reloads, and it very much needs to be._

**Comment by Myndex** [comment]
# Crashing again
This time the work I lost was updating a Gist in the Gist editor. Same cause here too it seems.

## Pretty please do something about this.
Just adding persistence in local storage would be enough for now!


Here's a crash log if its any help:

```

Process:               com.apple.WebKit.WebContent [67824]
Path:                  /Library/Apple/*/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent
Identifier:            com.apple.WebKit.WebContent
Version:               15613 (15613.3.9.1.16)
Build Info:            WebKit-7613003009001016~27 (613G51)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Safari [24662]
User ID:               501

Date/Time:             2023-10-08 06:02:23.729 -0700
OS Version:            Mac OS X 10.15.7 (19H2026)
Report Version:        12

Time Awake Since Boot: 470000 seconds
Time Since Wake:       7000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000018
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [67824]

VM Regions Near 0x18:
--> 
    __TEXT                 0000000100df9000-0000000100dfa000 [    4K] r-x/r-x SM=COW  /Library/Apple/*/WebKit.framework/Versions/A/XPCServices/com.apple.WebKit.WebContent.xpc/Contents/MacOS/com.apple.WebKit.WebContent

Application Specific Information:
Bundle controller class:
BrowserBundleController


Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   com.apple.WebCore             	0x000000046ac682db WebCore::characterCount(WebCore::SimpleRange const&, WTF::OptionSet<WebCore::TextIteratorBehavior>) + 27
1   com.apple.WebCore             	0x000000046ac6a82c WebCore::TextCheckingHelper::findFirstMisspelledWordOrUngrammaticalPhrase(bool) const + 956
2   com.apple.WebCore             	0x000000046ac077e8 WebCore::Editor::advanceToNextMisspelling(bool) + 7384
3   com.apple.WebKit              	0x000000046878ad9a WebKit::WebPage::advanceToNextMisspelling(bool) + 70
4   com.apple.WebKit              	0x00000004687a6640 WebKit::WebPage::didReceiveWebPageMessage(IPC::Connection&, IPC::Decoder&) + 39306
5   com.apple.WebKit              	0x00000004683552c2 IPC::MessageReceiverMap::dispatchMessage(IPC::Connection&, IPC::Decoder&) + 192
6   com.apple.WebKit              	0x00000004685cdf44 WebKit::WebProcess::didReceiveMessage(IPC::Connection&, IPC::Decoder&) + 28
7   com.apple.WebKit              	0x0000000468344ba4 IPC::Connection::dispatchMessage(std::__1::unique_ptr<IPC::Decoder, std::__1::default_delete<IPC::Decoder> >) + 594
8   com.apple.WebKit              	0x0000000468347329 WTF::Detail::CallableWrapper<IPC::Connection::enqueueIncomingMessage(std::__1::unique_ptr<IPC::Decoder, std::__1::default_delete<IPC::Decoder> >)::$_13, void>::call() + 183
9   com.apple.JavaScriptCore      	0x000000046db9c181 WTF::RunLoop::performWork() + 481
10  com.apple.JavaScriptCore      	0x000000046db9cc1a WTF::RunLoop::performWork(void*) + 26
11  com.apple.CoreFoundation      	0x00007fff354c6b64 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
12  com.apple.CoreFoundation      	0x00007fff354c6b03 __CFRunLoopDoSource0 + 103
13  com.apple.CoreFoundation      	0x00007fff354c691d __CFRunLoopDoSources0 + 209
14  com.apple.CoreFoundation      	0x00007fff354c5639 __CFRunLoopRun + 937
15  com.apple.CoreFoundation      	0x00007fff354c4c33 CFRunLoopRunSpecific + 466
16  com.apple.Foundation          	0x00007fff37b87c48 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 212
17  com.apple.Foundation          	0x00007fff37c3a6ef -[NSRunLoop(NSRunLoop) run] + 76
18  libxpc.dylib                  	0x00007fff6e84c4ea _xpc_objc_main.cold.4 + 49
19  libxpc.dylib                  	0x00007fff6e84c430 _xpc_objc_main + 559
20  libxpc.dylib                  	0x00007fff6e84bf63 xpc_main + 377
21  com.apple.WebKit              	0x00000004681c143b WebKit::XPCServiceMain(int, char const**) + 266
22  libdyld.dylib                 	0x00007fff6e5facc9 start + 1

Thread 1:: JavaScriptCore libpas scavenger
0   libsystem_kernel.dylib        	0x00007fff6e73e872 __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff6e803425 _pthread_cond_wait + 698
2   com.apple.JavaScriptCore      	0x000000046dc3d05f scavenger_thread_main + 1359
3   libsystem_pthread.dylib       	0x00007fff6e803109 _pthread_start + 148
4   libsystem_pthread.dylib       	0x00007fff6e7feb8b thread_start + 15

Thread 2:: WebCore: Scrolling
0   libsystem_kernel.dylib        	0x00007fff6e73bdea mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff6e73c160 mach_msg + 60
2   com.apple.CoreFoundation      	0x00007fff354c6d07 __CFRunLoopServiceMachPort + 247
3   com.apple.CoreFoundation      	0x00007fff354c57c1 __CFRunLoopRun + 1329
4   com.apple.CoreFoundation      	0x00007fff354c4c33 CFRunLoopRunSpecific + 466
5   com.apple.CoreFoundation      	0x00007fff3554d33c CFRunLoopRun + 40
6   com.apple.JavaScriptCore      	0x000000046db9d0a3 WTF::RunLoop::run() + 19
7   com.apple.JavaScriptCore      	0x000000046dbb3800 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*) + 256
8   com.apple.JavaScriptCore      	0x000000046d9ab8d9 WTF::wtfThreadEntryPoint(void*) + 9
9   libsystem_pthread.dylib       	0x00007fff6e803109 _pthread_start + 148
10  libsystem_pthread.dylib       	0x00007fff6e7feb8b thread_start + 15

Thread 3:
0   libsystem_pthread.dylib       	0x00007fff6e7feb68 start_wqthread + 0

Thread 4:
0   libsystem_pthread.dylib       	0x00007fff6e7feb68 start_wqthread + 0

Thread 5:
0   libsystem_pthread.dylib       	0x00007fff6e7feb68 start_wqthread + 0

Thread 6:: JIT Worklist Helper Thread
0   libsystem_kernel.dylib        	0x00007fff6e73e872 __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff6e803425 _pthread_cond_wait + 698
2   com.apple.JavaScriptCore      	0x000000046db96dd6 WTF::ParkingLot::parkConditionallyImpl(void const*, WTF::ScopedLambda<bool ()> const&, WTF::ScopedLambda<void ()> const&, WTF::TimeWithDynamicClockType const&) + 2470
3   com.apple.JavaScriptCore      	0x000000046db65cfa bool WTF::Condition::waitUntilUnchecked<WTF::Lock>(WTF::Lock&, WTF::TimeWithDynamicClockType const&) + 154
4   com.apple.JavaScriptCore      	0x000000046db660cf WTF::Detail::CallableWrapper<WTF::AutomaticThread::start(WTF::AbstractLocker const&)::$_0, void>::call() + 319
5   com.apple.JavaScriptCore      	0x000000046dbb3800 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*) + 256
6   com.apple.JavaScriptCore      	0x000000046d9ab8d9 WTF::wtfThreadEntryPoint(void*) + 9
7   libsystem_pthread.dylib       	0x00007fff6e803109 _pthread_start + 148
8   libsystem_pthread.dylib       	0x00007fff6e7feb8b thread_start + 15

Thread 7:
0   libsystem_pthread.dylib       	0x00007fff6e7feb68 start_wqthread + 0

Thread 0 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000045  rbx: 0x0000000000000000  rcx: 0x0000000000000049  rdx: 0x0000000000000000
  rdi: 0x00007ffeeee049e0  rsi: 0x0000000000000000  rbp: 0x00007ffeeee04870  rsp: 0x00007ffeeee04720
   r8: 0x000000046f140010   r9: 0x00000004965931b8  r10: 0x0000000000000001  r11: 0x00000004780ac0f0
  r12: 0x0000000000000000  r13: 0x00007ffeeee04d00  r14: 0x00000004964b0b80  r15: 0x00000004964b0b80
  rip: 0x000000046ac682db  rfl: 0x0000000000010202  cr2: 0x0000000000000018
  
Logical CPU:     6
Error Code:      0x00000006 (no mapping for user data write)
Trap Number:     14

```


   #crashing #github #editor #lostwork #pleasehelp

**Comment by MasterInQuestion** [comment]
Try "github.dev", a variant of "vscode.dev" (Visual Studio Code Web).

    [ Quote Myndex @ CE 2023-11-01 07:26:43 UTC:
[https://github.com/orgs/community/discussions/73118#discussion-5800465](https://github.com/orgs/community/discussions/73118#discussion-5800465)
    I am blind [ low vision ] and ... ]
<^>    Er... how do you review the accessibility of things?

    [ Quote Myndex @ CE 2023-09-14 12:08:51 UTC:
[https://github.com/orgs/community/discussions/67089#discussion-5627157](https://github.com/orgs/community/discussions/67089#discussion-5627157)
    I am disabled. Due to my low vision and other issues, I can easily and accidentally click someplace on the UI to cause navigation to a new page. ]

    ----

    You are using a rather outdated version of browser (Safari 15.6.1, released 20220818).
    And OS too (discontinued 20221130):
    https://en.wikipedia.org/wiki/MacOS_Catalina#Security
    .
    Both have been End of Life.

    2 factors combined, you are very susceptible to cyber attacks.
    Your system may be easily pwned during web browsing (directed to malicious page, or unconsciously loading malicious content from legitimate page etc.).

    https://github.com/orgs/community/discussions/67089#discussioncomment-7225511
<^>    Such browser crashes are notable symbols of serious exploitable flaws.
\
\
    Current infrastructures don't tend to be well designed before deployment:
    Thus the continuous patching and patching, fixing one thing breaking another etc. ...

    Modern websites may also selectively drop support for old browsers:
    Due to the development burden caused by previously poorly defined interfaces.
    So... Good luck.

**Reply by Myndex** [reply]
Literally all of my hardware is legacy, and I am disabled/retired. When I get my pension issues straightened out I'll buy something newer. Until them I'm stuck.

And herein lies the actual problem: there is no aspect of this Macbook, nor the dozen or so MacPros I have, that make them nonservicable--they are all still perfectly useful—except for the removal of support from apple. They are fast, have ample RAM, but it's Apple's dropping support that is making them (me) SOL. 

It's not a matter up upgrading software, it's a matter of capital investment for something I don't need other than not being able to have the latest set of little ones and zeros...

**Reply by MasterInQuestion** [reply]
Indeed not a matter of upgrading software.
    It's a matter of general implementation of all things.

    The recent CVE-2023-5129, CVE-2023-5217 shall be applicable to you.
    As a workaround, consider air-gap your sensitive data.

**Reply by Myndex** [reply]
Thank you... my actual research data and the protocol software is on airgapped machines. Everything else is behind a firewall.

Interestingly, this version of Safari does not recognize WebP images at all. Chrome is up to date...

**Reply by MasterInQuestion** [reply]
So you happened to have avoided which...

    See also:
    https://caniuse.com/webp
